File: app/src/main/java/org/nervos/neuron/activity/transactionlist/view/TransactionListActivity.java
package org.nervos.neuron.activity.transactionlist.view;

import android.content.Intent;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.AppCompatButton;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;
import org.nervos.neuron.R;
import org.nervos.neuron.activity.NBaseActivity;
import org.nervos.neuron.activity.ReceiveQrCodeActivity;
import org.nervos.neuron.activity.TransactionDetailActivity;
import org.nervos.neuron.activity.transactionlist.model.TransactionAdapter;
import org.nervos.neuron.activity.transactionlist.presenter.TransactionListPresenter;
import org.nervos.neuron.activity.transfer.TransferActivity;
import org.nervos.neuron.item.TokenItem;
import org.nervos.neuron.item.WalletItem;
import org.nervos.neuron.item.transaction.TransactionResponse;
import org.nervos.neuron.util.db.DBWalletUtil;
import org.nervos.neuron.view.TitleBar;
import org.nervos.neuron.view.TokenProfileView;
import org.nervos.neuron.view.loadmore.OnLoadMoreListener;
import org.nervos.neuron.view.loadmore.RecyclerViewLoadMoreScroll;

import java.util.ArrayList;
import java.util.List;

public class TransactionListActivity extends NBaseActivity {

    public static final String TRANSACTION_TOKEN = "TRANSACTION_TOKEN";
    public static final String TRANSACTION_STATUS = "TRANSACTION_STATUS";

    private List<TransactionResponse> transactionResponseList = new ArrayList<>();
    private WalletItem walletItem;

    private RecyclerView recyclerView;
    private SwipeRefreshLayout swipeRefreshLayout;
    private TextView tvTokenWarning;
    private TitleBar titleBar;
    private AppCompatButton receiveButton, transferButton;
    private TokenProfileView mTokenProfile;

    private TokenItem tokenItem;
    private TransactionAdapter transactionAdapter;
    private RecyclerViewLoadMoreScroll scrollListener;
    private TransactionListPresenter presenter;

    private int mPage = 0;

    @Override
    protected int getContentLayout() {
        return R.layout.activity_transaction_list;
    }

    @Override
    protected void initView() {
        recyclerView = findViewById(R.id.transaction_recycler);
        swipeRefreshLayout = findViewById(R.id.swipe_refresh_layout);
        receiveButton = findViewById(R.id.receive_token);
        transferButton = findViewById(R.id.transfer_token);
        titleBar = findViewById(R.id.title);
        tvTokenWarning = findViewById(R.id.tv_token_warning);
        mTokenProfile = findViewById(R.id.view_token_profile);
    }

    @Override
    protected void initData() {
        walletItem = DBWalletUtil.getCurrentWallet(mActivity);
        tokenItem = getIntent().getParcelableExtra(TRANSACTION_TOKEN);
        titleBar.setTitle(tokenItem.symbol);
        tvTokenWarning.setVisibility(isTestToken() ? View.VISIBLE : View.GONE);

        presenter = new TransactionListPresenter(this, tokenItem, listener);
        mTokenProfile.init(tokenItem);
        initTransactionData();
    }

    private void initTransactionData() {
        initAdapter();
        showProgressBar();
        presenter.getTransactionList(mPage);
    }

    @Override
    protected void initAction() {
        receiveButton.setOnClickListener(v -> startActivity(new Intent(mActivity, ReceiveQrCodeActivity.class)));
        transferButton.setOnClickListener(v -> {
            Intent intent = new Intent(mActivity, TransferActivity.class);
            intent.putExtra(TransferActivity.EXTRA_TOKEN, tokenItem);
            startActivity(intent);
        });
        swipeRefreshLayout.setOnRefreshListener(() -> {
            mPage = 0;
            presenter.getTransactionList(mPage);
        });
    }

    private void initAdapter() {
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(mActivity);
        linearLayoutManager.setOrientation(LinearLayoutManager.VERTICAL);
        recyclerView.setLayoutManager(linearLayoutManager);
        transactionAdapter = new TransactionAdapter(this, transactionResponseList, walletItem.address);
        recyclerView.setAdapter(transactionAdapter);

        scrollListener = new RecyclerViewLoadMoreScroll(linearLayoutManager);
        scrollListener.setOnLoadMoreListener(new OnLoadMoreListener() {
            @Override
            public void onLoadMore() {
                LoadMoreData();
            }
        });

        recyclerView.addOnScrollListener(scrollListener);

        transactionAdapter.setOnItemClickListener((view, position) -> {
            TransactionResponse response = transactionResponseList.get(position);
            Intent intent = new Intent(mActivity, TransactionDetailActivity.class);
            intent.putExtra(TransactionDetailActivity.TRANSACTION_DETAIL, response);
            intent.putExtra(TRANSACTION_TOKEN, tokenItem);
            intent.putExtra(TRANSACTION_STATUS, response.status);
            startActivity(intent);
        });
    }

    private void LoadMoreData() {
        transactionAdapter.addLoadingView();
        presenter.getTransactionList(mPage);
    }

    private TransactionListPresenter.TransactionListPresenterImpl listener
            = new TransactionListPresenter.TransactionListPresenterImpl() {
        @Override
        public void hideProgressBar() {
            swipeRefreshLayout.post(() -> dismissProgressBar());
        }

        @Override
        public void setRefreshing(boolean refreshing) {
            swipeRefreshLayout.setRefreshing(refreshing);
        }

        @Override
        public void updateNewList(List<TransactionResponse> list) {
            mPage++;
            transactionResponseList = list;
            transactionAdapter.refresh(transactionResponseList);
        }

        @Override
        public void refreshList(List<TransactionResponse> list) {
            mPage++;
            transactionAdapter.removeLoadingView();
            transactionResponseList.addAll(list);
            transactionAdapter.refresh(transactionResponseList);
            scrollListener.setLoaded();
        }

        @Override
        public void noMoreLoading() {
            transactionAdapter.removeLoadingView();
            scrollListener.setLoaded();
            if (transactionResponseList.size() > 0) {
                Toast.makeText(mActivity, R.string.no_more_transaction_data, Toast.LENGTH_SHORT).show();
            }
        }
    };

    private boolean isTestToken() {
        return "NATT".equalsIgnoreCase(tokenItem.symbol) || "MBA".equalsIgnoreCase(tokenItem.symbol);
    }

}

File: app/src/main/java/org/nervos/neuron/activity/transactionlist/view/TransactionListActivity.kt
package org.nervos.neuron.activity.transactionlist.view

import android.content.Intent
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import kotlinx.android.synthetic.main.activity_transaction_list.*
import org.nervos.neuron.R
import org.nervos.neuron.activity.NBaseActivity
import org.nervos.neuron.activity.ReceiveQrCodeActivity
import org.nervos.neuron.activity.SimpleWebActivity
import org.nervos.neuron.activity.TransactionDetailActivity
import org.nervos.neuron.activity.transactionlist.model.TransactionAdapter
import org.nervos.neuron.activity.transactionlist.presenter.TransactionListPresenter
import org.nervos.neuron.activity.transfer.TransferActivity
import org.nervos.neuron.item.EthErc20TokenInfoItem
import org.nervos.neuron.item.TokenItem
import org.nervos.neuron.item.TransactionItem
import org.nervos.neuron.item.WalletItem
import org.nervos.neuron.service.http.HttpUrls
import org.nervos.neuron.util.AddressUtil
import org.nervos.neuron.util.ConstUtil
import org.nervos.neuron.util.db.DBWalletUtil
import org.nervos.neuron.view.TitleBar
import org.web3j.crypto.Keys
import java.math.RoundingMode
import java.text.DecimalFormat
import java.util.*

class TransactionListActivity : NBaseActivity() {

    private var transactionItemList: List<TransactionItem> = ArrayList()
    private var walletItem: WalletItem? = null

    private var titleBar: TitleBar? = null
    private var tokenItem: TokenItem? = null

    private var transactionAdapter: TransactionAdapter? = null

    private var presenter: TransactionListPresenter? = null

    private var describe: String? = null

    private val listener = object : TransactionListPresenter.TransactionListPresenterImpl {
        override fun hideProgressBar() {
            swipe_refresh_layout.post { dismissProgressBar() }
        }

        override fun setRefreshing(refreshing: Boolean) {
            swipe_refresh_layout.isRefreshing = refreshing
        }

        override fun refreshList(list: List<TransactionItem>) {
            transactionItemList = list
            transactionAdapter!!.refresh(transactionItemList)
        }

        override fun getTokenDescribe(item: EthErc20TokenInfoItem) {
            describe = item.overView.zh
            tv_token_des_first.post {
                tv_token_symbol.text = item.symbol
                tv_token_des_first.text = item.overView.zh
                setDesSecondLine()
            }
        }

        override fun showTokenDescribe(show: Boolean) {
            swipe_refresh_layout.post {
                if (show) {
                    cl_token_des.visibility = View.VISIBLE
                    presenter!!.setTokenLogo(iv_token_logo)
                    initBalance()
                    var address = tokenItem!!.contractAddress
                    if (AddressUtil.isAddressValid(address))
                        address = Keys.toChecksumAddress(address)
                    val finalAddress = address
                    cl_token_des.setOnClickListener { SimpleWebActivity.gotoSimpleWeb(mActivity, HttpUrls.TOKEN_ERC20_DETAIL.replace("@address", finalAddress)) }
                } else
                    cl_token_des.visibility = View.GONE
            }
        }

        override fun getCurrency(currency: String) {
            tv_balance.post { tv_balance.text = currency }
        }
    }

    private val isTestToken: Boolean
        get() = "NATT".equals(tokenItem!!.symbol, ignoreCase = true) || "MBA".equals(tokenItem!!.symbol, ignoreCase = true)

    override fun getContentLayout(): Int {
        return R.layout.activity_transaction_list
    }

    override fun initView() {
        titleBar = findViewById(R.id.title)
    }

    override fun initData() {
        walletItem = DBWalletUtil.getCurrentWallet(mActivity)
        tokenItem = intent.getParcelableExtra(TRANSACTION_TOKEN)
        titleBar!!.title = tokenItem!!.symbol
        showProgressBar()
        presenter = TransactionListPresenter(this, tokenItem, listener)
        initAdapter()
        presenter!!.getTransactionList(walletItem!!.address)
        initDescribe()
        val formater = DecimalFormat("0.####")
        formater.roundingMode = RoundingMode.FLOOR

        tv_token_warning.visibility = if (isTestToken) View.VISIBLE else View.GONE
    }

    override fun initAction() {
        receive_token.setOnClickListener { startActivity(Intent(mActivity, ReceiveQrCodeActivity::class.java)) }
        transfer_token.setOnClickListener {
            val intent = Intent(mActivity, TransferActivity::class.java)
            intent.putExtra(TransferActivity.EXTRA_TOKEN, tokenItem)
            startActivity(intent)
        }
        swipe_refresh_layout.setOnRefreshListener { presenter!!.getTransactionList(walletItem!!.address) }
    }

    private fun initAdapter() {
        transaction_recycler.layoutManager = LinearLayoutManager(mActivity)
        transactionAdapter = TransactionAdapter(this, transactionItemList, walletItem!!.address)
        transaction_recycler.adapter = transactionAdapter

        transactionAdapter!!.setOnItemClickListener { _, position ->
            val item = transactionItemList[position]
            if (item.status != TransactionItem.PENDING) {
                val intent = Intent(mActivity, TransactionDetailActivity::class.java)
                intent.putExtra(TransactionDetailActivity.TRANSACTION_DETAIL, item)
                intent.putExtra(TRANSACTION_TOKEN, tokenItem)
                startActivity(intent)
            }
        }
    }

    private fun initDescribe() {
        if (presenter!!.isEthereum(tokenItem)) {
            if (!presenter!!.isNativeToken(tokenItem)) {
                presenter!!.getTokenDescribe()
            } else {
                cl_token_des.visibility = View.VISIBLE
                describe = resources.getString(R.string.ETH_Describe)
                tv_token_des_first.setText(R.string.ETH_Describe)
                tv_token_symbol.text = ConstUtil.ETH
                setDesSecondLine()
                presenter!!.getBalance()
                cl_token_des.setOnClickListener { SimpleWebActivity.gotoSimpleWeb(mActivity, HttpUrls.TOKEN_DETAIL.replace("@address", "ethereum")) }
            }
        }
    }

    private fun initBalance() {
        if (tokenItem!!.balance != 0.0 && presenter!!.isEthereum(tokenItem)) {
            presenter!!.getBalance()
        }
    }

    private fun setDesSecondLine() {
        tv_token_des_first.postDelayed({
            val layout = tv_token_des_first.layout
            val mDesText = tv_token_des_first.layout.text.toString()
            val srcStr = StringBuilder(mDesText)
            val lineStr = srcStr.subSequence(layout.getLineStart(0), layout.getLineEnd(0))
                    .toString()
            val length = lineStr.length
            if (length > 0 && describe!!.length > length
                    && tv_token_des_second.text.toString().isEmpty()) {
                val secondText: String
                if ((length * 1.5).toInt() > describe!!.length) {
                    secondText = describe!!.substring(length)
                } else {
                    secondText = describe!!.substring(length, (length * 1.5).toInt()) + "..."
                }
                tv_token_des_second.text = secondText
            }
        }, 300)
    }

    companion object {

        val TRANSACTION_TOKEN = "TRANSACTION_TOKEN"
    }

}
